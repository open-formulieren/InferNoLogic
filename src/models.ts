// Copyright (c) 2023 Adam Jones
// Copyright (c) 2023 Maykin
//
// SPDX-License-Identifier: MIT

// Expressions

// e ::= x
//     | e1 e2
//     | \x -> e
//     | let x = e1 in e2

export type Expression =
  | VariableExpression
  | ApplicationExpression
  | AbstractionExpression
  | LetExpression
  | NumberLiteral
  | StringLiteral;

export interface NumberLiteral {
  type: 'num';
  x: number;
}

export interface StringLiteral {
  type: 'str';
  x: string;
}

export interface VariableExpression {
  type: 'var';
  x: string;
}

export interface ApplicationExpression {
  type: 'app';
  e1: Expression;
  e2: Expression;
}

export interface AbstractionExpression {
  type: 'abs';
  x: string;
  e: Expression;
}

export interface LetExpression {
  type: 'let';
  x: string;
  e1: Expression;
  e2: Expression;
}

// Types

// mu ::= a
//      | C mu_0 ... mu_n

// sigma ::= mu
//         | Va. sigma

export type MonoType = TypeVariable | TypeFunctionApplication;

export type PolyType = MonoType | TypeQuantifier;

export type TypeFunction = '->' | 'Array' | 'Boolean' | 'Number' | 'String' | 'Null';

export interface TypeVariable {
  type: 'ty-var';
  a: string;
  explain?: [MonoType, TypeExplanation]; // Optional explanation
}

export type TypeExplanation =
  | {type: 'ExplainAlias'; path1: ExplainPath; path2: ExplainPath; expr: Expression}
  | {type: 'ExplainInstan'; path: ExplainPath; expr: Expression};

export type ExplainPath = TypeVariable[];

export interface TypeFunctionApplication {
  type: 'ty-app';
  C: TypeFunction;
  mus: MonoType[];
}

export interface TypeQuantifier {
  type: 'ty-quantifier';
  a: string;
  sigma: PolyType;
}

// Contexts

export const ContextMarker = Symbol();
export type Context = {[ContextMarker]: boolean; [variable: string]: PolyType};

export const makeContext = (raw: {
  [ContextMarker]?: boolean;
  [variable: string]: PolyType;
}): Context => {
  raw[ContextMarker] = true;
  return raw as Context;
};

export const isContext = (something: unknown): something is Context => {
  return typeof something === 'object' && something !== null && ContextMarker in something;
};
