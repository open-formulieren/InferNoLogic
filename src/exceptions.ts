export class InferenceError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InferenceError';
  }
}
