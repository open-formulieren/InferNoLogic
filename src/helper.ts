// Copyright (c) 2023 Adam Jones
//
// SPDX-License-Identifier: MIT
import {
  Context,
  ExplainPath,
  Expression,
  MonoType,
  PolyType,
  TypeFunctionApplication,
  TypeVariable,
  isContext,
  makeContext,
} from './models';

// substitutions

export type Substitution = {
  type: 'substitution';
  (m: MonoType): MonoType;
  (t: PolyType): PolyType;
  (c: Context): Context;
  (s: Substitution): Substitution;
  raw: {[typeVariables: string]: MonoType};
};

export const makeSubstitution = (raw: Substitution['raw']): Substitution => {
  const fn = ((arg: MonoType | PolyType | Context | Substitution) => {
    if (arg.type === 'substitution') return combine(fn, arg);
    return apply(fn, arg);
  }) as Substitution;
  fn.type = 'substitution';
  fn.raw = raw;
  return fn;
};

function apply<T extends MonoType | PolyType | Context>(substitution: Substitution, value: T): T;
function apply(
  s: Substitution,
  value: MonoType | PolyType | Context
): MonoType | PolyType | Context {
  if (isContext(value)) {
    return makeContext(Object.fromEntries(Object.entries(value).map(([k, v]) => [k, apply(s, v)])));
  }

  if (value.type === 'ty-var') {
    if (s.raw[value.a]) return s.raw[value.a];
    return value;
  }

  if (value.type === 'ty-app') {
    return {...value, mus: value.mus.map(m => apply(s, m))};
  }

  if (value.type === 'ty-quantifier') {
    return {...value, sigma: apply(s, value.sigma)};
  }
  ((_: never): never => {
    throw new Error('Unknown argument passed to substitution');
  })(value);
}

const combine = (s1: Substitution, s2: Substitution): Substitution => {
  return makeSubstitution({
    ...s1.raw,
    ...Object.fromEntries(Object.entries(s2.raw).map(([k, v]) => [k, s1(v)])),
  });
};

// new type variable
let currentTypeVar = 0;
export const newTypeVar = (): TypeVariable => ({
  type: 'ty-var',
  a: `t${currentTypeVar++}`,
});

// instantiate
// mappings = { a |-> t0, b |-> t1 }
// Va. Vb. a -> b
// t0 -> t1
export const instantiate = (
  type: PolyType,
  mappings: Map<string, TypeVariable> = new Map()
): MonoType => {
  if (type.type === 'ty-var') {
    return mappings.get(type.a) ?? type;
  }

  if (type.type === 'ty-app') {
    return {...type, mus: type.mus.map(m => instantiate(m, mappings))};
  }

  if (type.type === 'ty-quantifier') {
    mappings.set(type.a, newTypeVar());
    return instantiate(type.sigma, mappings);
  }

  ((_: never): never => {
    throw new Error('Unknown type passed to instantiate');
  })(type);
};

// generalise
export const generalise = (ctx: Context, type: MonoType): PolyType => {
  const quantifiers = diff(freeVars(type), freeVars(ctx));
  let t: PolyType = type;
  quantifiers.forEach(q => {
    t = {type: 'ty-quantifier', a: q, sigma: t};
  });
  return t;
};

const diff = <T>(a: T[], b: T[]): T[] => {
  const bset = new Set(b);
  return a.filter(v => !bset.has(v));
};

const freeVars = (value: PolyType | Context): string[] => {
  if (isContext(value)) {
    return Object.values(value).flatMap(freeVars);
  }

  if (value.type === 'ty-var') {
    return [value.a];
  }

  if (value.type === 'ty-app') {
    return value.mus.flatMap(freeVars);
  }

  if (value.type === 'ty-quantifier') {
    return freeVars(value.sigma).filter(v => v !== value.a);
  }

  ((_: never): never => {
    throw new Error('Unknown argument passed to substitution');
  })(value);
};

// unify

export const unify = (
  type1: MonoType,
  type2: MonoType,
  expr: Expression,
  path1: ExplainPath = [],
  path2: ExplainPath = []
): Substitution => {
  if (type1.type === 'ty-var' && type2.type === 'ty-var' && type1.a === type2.a) {
    return makeSubstitution({});
  }

  if (type1.type === 'ty-var') {
    if (contains(type2, type1))
      throw new Error(`Infinite type detected: ${type1} occurs in ${type2}`);

    if (type2.type === 'ty-var') {
      // var with other name -> explain
      // TODO? reverseAliasPath(type1);
      type1.explain = [type2, {type: 'ExplainAlias', path1, path2, expr}];
    } else if (type2.type === 'ty-app') {
      // instantiation
      // TODO? reverseAliasPath(type1);
      type1.explain = [type2, {type: 'ExplainInstan', path: path1, expr}];
    } else {
      ((_: never): never => {
        throw new Error('Unknown argument passed to unify');
      })(type2);
    }
    return makeSubstitution({
      [type1.a]: type2,
    });
  }

  if (type2.type === 'ty-var') {
    return unify(type2, type1, expr, path2, path1);
  }

  if (type1.C !== type2.C) {
    const msg = formatUnificationError(type1, type2, expr, path1, path2);
    throw new Error(msg);
  }

  if (type1.mus.length !== type2.mus.length) {
    throw new Error(`Could not unify types (different argument lengths): ${type1} and ${type2}`);
  }

  let s: Substitution = makeSubstitution({});
  for (let i = 0; i < type1.mus.length; i++) {
    s = s(
      unify(s(type1.mus[i]), s(type2.mus[i]), expr, addHist(path1, type1), addHist(path2, type2))
    );
  }
  return s;
};

const formatUnificationError = (
  type1: TypeFunctionApplication,
  type2: TypeFunctionApplication,
  expr: Expression,
  _path1: ExplainPath,
  _path2: ExplainPath
): string => {
  // console.dir({type1, type2, expr, _path1, _path2}, {depth: Infinity});
  if (expr.type === 'app') {
    const msg = `"${reprExpression(expr.e1)}" expects "${reprExpression(expr.e2)}" to be a ${
      type1.C
    }, but it is a ${type2.C}`;
    return msg;
  }
  throw new Error(`Unexpected expression type ${expr}`);
};

export const reprExpression = (expr: Expression): string => {
  if (expr.type === 'app') return reprExpression(expr.e1);
  if (expr.type === 'var') return expr.x.startsWith('var: ') ? expr.x.substring(5) : expr.x;
  if (expr.type === 'num') return expr.x.toString();
  if (expr.type === 'str') return expr.x.toString();
  if (expr.type === 'abs') return `{"${expr.x}": ${reprExpression(expr.e)}}`;
  if (expr.type === 'let')
    return `"${expr.x}" = ${reprExpression(expr.e1)} in ${reprExpression(expr.e2)}`;
  ((_: never): never => {
    throw new Error(`Unexpected expression type ${expr}`);
  })(expr);
};

const addHist = (history: ExplainPath, root: MonoType): ExplainPath => {
  const _addHist = (root: MonoType): ExplainPath => {
    if (root.type === 'ty-app' || !root.explain) return [];
    const [ty, _explain] = root.explain;
    const explPath = _addHist(ty);
    // root.explain = [explPath, _explain]
    return [root, ...explPath];
  };
  // and return the new history
  return history.concat(_addHist(root));
};

const contains = (value: MonoType, type2: TypeVariable): boolean => {
  if (value.type === 'ty-var') {
    return value.a === type2.a;
  }

  if (value.type === 'ty-app') {
    return value.mus.some(t => contains(t, type2));
  }

  ((_: never): never => {
    throw new Error('Unknown argument passed to substitution');
  })(value);
};
