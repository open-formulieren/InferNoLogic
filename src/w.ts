// Copyright (c) 2023 Adam Jones
// Copyright (c) 2023 Maykin
//
// SPDX-License-Identifier: MIT
import {InferenceError} from './exceptions';
import {Substitution, generalise, instantiate, makeSubstitution, newTypeVar, unify} from './helper';
import {Context, Expression, MonoType, PolyType, makeContext} from './models';

export const W = (typEnv: Context, expr: Expression): [Substitution, MonoType] => {
  if (expr.type === 'var') {
    const value = typEnv[expr.x];
    if (value === undefined) throw new InferenceError(`Undefined variable: ${expr.x}`);
    return [makeSubstitution({}), instantiate(value)];
  }

  if (expr.type === 'num') {
    return [makeSubstitution({}), {type: 'ty-app', C: 'Number', mus: []}];
  }

  if (expr.type === 'str') {
    return [makeSubstitution({}), {type: 'ty-app', C: 'String', mus: []}];
  }

  if (expr.type === 'abs') {
    const beta = newTypeVar();
    const [s1, t1] = W(
      makeContext({
        ...typEnv,
        [expr.x]: beta,
      }),
      expr.e
    );
    return [
      s1,
      s1({
        type: 'ty-app',
        C: '->',
        mus: [beta, t1],
      }),
    ];
  }

  if (expr.type === 'app') {
    const [s1, t1] = W(typEnv, expr.e1);
    const [s2, t2] = W(s1(typEnv), expr.e2);
    const beta = newTypeVar();

    try {
      const s3 = unify(
        s2(t1),
        {
          type: 'ty-app',
          C: '->',
          mus: [t2, beta],
        },
        expr
      );
      return [s3(s2(s1)), s3(beta)];
    } catch (error) {
      const hasExplanation = ([k, t]: [string, PolyType]): boolean =>
        k.startsWith('var') && t.type == 'ty-var' && t.explain != undefined;
      const withExpl = Object.fromEntries(Object.entries(typEnv).filter(hasExplanation));
      withExpl;
      // TODO expr here !== expr from the throw site in unify!!
      // reversing explainPaths is tricky... stringifying the larger expr and the sub expr from unify
      // will probably provide enough context.
      throw error;
    }
  }

  if (expr.type === 'let') {
    const [s1, t1] = W(typEnv, expr.e1);
    const [s2, t2] = W(
      makeContext({
        ...s1(typEnv),
        [expr.x]: generalise(typEnv, t1),
      }),
      expr.e2
    );
    return [s2(s1), t2];
  }
  throw new InferenceError('Unknown expression type');
};
