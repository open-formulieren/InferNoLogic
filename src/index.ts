import type {Substitution} from './helper';
import type {Context} from './models';
import type {JSONValue} from './parser';
import {defaultContext, parseContext, parseJsonLogicExpression} from './parser';
import {W} from './w';

export {defaultContext};

interface InferenceResult {
  resultType: string;
  intermediateVariables: Substitution['raw'];
}

/**
 * @beta
 * @param jsonLogic - JsonLogic expression/rule
 * @param data - JsonLogic {@link https://jsonlogic.com/#data-driven | data} object
 * @param context - TypeEnv / Context / Î“ , you only need this is you've used {@link https://jsonlogic.com/add_operation.html | add_operation}
 * @returns a description of the
 */
export const infer = (
  jsonLogic: JSONValue,
  data: JSONValue,
  context: Context = defaultContext
): InferenceResult => {
  const typeenv = parseContext(data, context);
  const [subsitution, t] = W(...parseJsonLogicExpression(jsonLogic, typeenv));
  const type: string = 'C' in t ? t.C : JSON.stringify([t.a, Object.keys(subsitution.raw)]);
  return {
    resultType: type,
    intermediateVariables: subsitution.raw,
  };
};
