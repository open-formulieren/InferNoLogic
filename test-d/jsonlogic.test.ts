import {Substitution, newTypeVar} from '../src/helper';
import {M} from '../src/m';
import {MonoType} from '../src/models';
import type {JSONValue} from '../src/parser';
import {parseContext, parseJsonLogicExpression} from '../src/parser';
import {W} from '../src/w';
import cases from './tests.json';

type Case = [rule: JSONValue, data: JSONValue, result: JSONValue];
// obvious cast; imported json
// filter "# commment string"
const isCase = (testCase: string | unknown[]): testCase is Case =>
  Array.isArray(testCase) && testCase.length == 3;
type ErrorCase = [rule: JSONValue, data: JSONValue, result: JSONValue, error: string];

const isErrorCase = (testCase: string | unknown[]): testCase is ErrorCase =>
  Array.isArray(testCase) && testCase.length == 4;

const inferResultType = (jsonLogic: JSONValue, data: JSONValue, use: 'W' | 'M' = 'W'): string => {
  const context = parseContext(data);
  let substitution: Substitution, t: MonoType;
  if (use === 'W') {
    [substitution, t] = W(...parseJsonLogicExpression(jsonLogic, context));
    return 'C' in t ? t.C : JSON.stringify([t.a, Object.keys(substitution.raw)]);
  } else if (use === 'M') {
    t = newTypeVar();
    substitution = M(...parseJsonLogicExpression(jsonLogic, context), t);
    t = substitution(t);
    return 'C' in t ? t.C : JSON.stringify([t.a, Object.keys(substitution.raw)]);
  }
  ((_: never): never => {
    throw Error();
  })(use);
};

const getType = (obj: unknown): 'Boolean' | 'Number' | 'String' | 'Array' | 'Null' | 'Object' =>
  typeof obj === 'boolean'
    ? 'Boolean'
    : obj === null
    ? 'Null'
    : typeof obj === 'number'
    ? 'Number'
    : typeof obj === 'string'
    ? 'String'
    : Array.isArray(obj)
    ? 'Array'
    : 'Object';

describe('Test against JsonLogic suite', () => {
  // https://jsonlogic.com/tests.json
  it.each(cases.filter(isCase))(
    'can infer %j, with context %j, to have the type of %j',
    (rule: JSONValue, data: JSONValue, result: JSONValue) => {
      const t = inferResultType(rule, data);
      expect(t).toBe(getType(result));
    }
  );
  it.each(cases.filter(isErrorCase))(
    'can detect %j, with context %j, does not typecheck',
    (rule: JSONValue, data: JSONValue, result: JSONValue, error: string) => {
      if (error.startsWith('Valid')) {
        // e.g. expressions like {"and": [1, 2]} are valid, but we infer them as Boolean
        // not Number (the type of the 2 literal)
        const t = inferResultType(rule, data);
        expect(t).not.toBe(getType(result));
      } else if (error == 'Snapshot')
        expect(() => inferResultType(rule, data)).toThrowErrorMatchingSnapshot();
      else if (error.startsWith('TODO')) expect(() => inferResultType(rule, data)).toThrow();
      else expect(() => inferResultType(rule, data)).toThrow(error);
    }
  );

  test('{"+": [{"var": "a"}, "1"]} with a: "1" ', () => {
    expect(() => inferResultType({'+': [{var: 'a'}, 1]}, {a: '1'})).toThrow(
      `"+" expects "a" to be a Number, but it is a String`
    );
  });
  test('{"+": [1, "1"]}', () => {
    expect(() => inferResultType({'+': ['1', 1]}, {})).toThrow(
      `"+" expects "1" to be a Number, but it is a String`
    );
  });
  test('{">": ["2", 1]}', () => {
    expect(() => inferResultType({'>': ['2', 1]}, {})).toThrowErrorMatchingInlineSnapshot(
      `"">" expects "2" to be a Number, but it is a String"`
    );
  });
  test('{"===": [{"var": ["items expression", [["value", "label"]]]}, {"var": "current_year"}]}', () => {
    expect(() =>
      inferResultType(
        {
          '===': [
            {
              if: [
                {'!!': [{var: 'items expression'}]},
                {var: 'items expression'},
                [['value', 'label']],
              ],
            },

            {var: 'current_year'},
          ],
        },

        // {'===': [{var: ['items expression', [['value', 'label']]]}, {var: 'current_year'}]},
        {current_year: 2023}
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `""===" expects "current_year" to be a Array, but it is a Number"`
    ); // TODO improve this. While correct, user can't change current_year
  });
});
