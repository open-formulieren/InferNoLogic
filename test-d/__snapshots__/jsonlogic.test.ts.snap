// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test against JsonLogic suite can detect {"!==":[1,"1"]}, with context {}, does not typecheck 1`] = `""!==" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"*":["1",1]}, with context {}, does not typecheck 1`] = `""*" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"+":["1",1]}, with context {}, does not typecheck 1`] = `""+" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"/":["1",1]}, with context {}, does not typecheck 1`] = `""/" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"<":["1",2]}, with context {}, does not typecheck 1`] = `""<" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"<=":["1",2]}, with context {}, does not typecheck 1`] = `""<=" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"===":[0,"0"]}, with context null, does not typecheck 1`] = `""===" expects "0" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"===":[1,"1"]}, with context {}, does not typecheck 1`] = `""===" expects "1" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {">":["2",1]}, with context {}, does not typecheck 1`] = `"">" expects "2" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {">=":["2",1]}, with context {}, does not typecheck 1`] = `"">=" expects "2" to be a Number, but it is a String"`;

exports[`Test against JsonLogic suite can detect {"-":["1",1]}, with context {}, does not typecheck 1`] = `""-" expects "1" to be a Number, but it is a String"`;
